packages/integrations/mdx/src/index.ts

```diff
@@ -132,11 +132,18 @@ export default function mdx(mdxOptions: MdxOptions = {}): AstroIntegration {
 								transform(code, id) {
 									if (!id.endsWith('.mdx')) return;
 
-									// Ensures styles and scripts are injected into a `<head>`
-									// When a layout is not applied
-									code += `\nMDXContent[Symbol.for('astro.needsHeadRendering')] = !Boolean(frontmatter.layout);`;
-
-									const [, moduleExports] = parseESM(code);
+									const [moduleImports, moduleExports] = parseESM(code);
+
+									// Fragment import should already be injected, but check just to be safe.
+									const importsFromJSXRuntime = moduleImports
+										.filter(({ n }) => n === 'astro/jsx-runtime')
+										.map(({ ss, se }) => code.substring(ss, se));
+									const hasFragmentImport = importsFromJSXRuntime.some((statement) =>
+										/[\s,{](Fragment,|Fragment\s*})/.test(statement)
+									);
+									if (!hasFragmentImport) {
+										code = 'import { Fragment } from "astro/jsx-runtime"\n' + code;
+									}
```

Comment:
Turns out the parsed imports aren’t in the friendliest of formats. This is pretty thorough and will inject the import if it doesn’t find one. If you think an error in that scenario would be more appropriate, let me know.